File: app/src/main/java/io/catter2/di/CachedRetrofitCatApiDIModule.java
package io.catter2.di;

import io.catter2.cat_api.CacheTheCatAPI;
import io.catter2.cat_api.RetrofitTheCatAPI;
import io.catter2.cat_api.TheCatAPI;

public class CachedRetrofitCatApiDIModule extends TheCatAPIDIModule {
    @Override
    public TheCatAPI provideTheCatAPI() {
        return provideTheCatAPICached(provideTheCatAPIRetrofit());
    }

    public TheCatAPI provideTheCatAPIRetrofit() {
        return new RetrofitTheCatAPI();
    }

    public TheCatAPI provideTheCatAPICached(TheCatAPI catAPI) {
        return new CacheTheCatAPI(catAPI);
    }
}

File: app/src/main/java/io/catter2/di/CachedRetrofitCatApiDIModule.kt
package io.catter2.di

import io.catter2.cat_api.CacheTheCatAPI
import io.catter2.cat_api.RetrofitTheCatAPI
import io.catter2.cat_api.TheCatAPI

open class CachedRetrofitCatApiDIModule : TheCatAPIDIModule() {
    override fun provideTheCatAPI(): TheCatAPI = provideTheCatAPICached(provideTheCatAPIRetrofit())
    protected fun provideTheCatAPIRetrofit(): TheCatAPI = RetrofitTheCatAPI()
    private fun provideTheCatAPICached(catAPI: TheCatAPI): TheCatAPI = CacheTheCatAPI(catAPI)
}
